File: golang/v2/pm.go
package layrzprotocol

import (
	"encoding/base64"
	"errors"
	"fmt"
	"strconv"
	"strings"
)

// Based on the `Layrz Protocol v2` specification. PmPacket is the media packet
// sent from the device to the server
type PmPacket struct {
	// Is the timestamp of the response packet
	Filename *string

	// Is the command id of the response packet
	ContentType *string

	// Is the message of the response packet
	Data *[]byte
}

// FromPacket is a method that converts a raw packet to a PaPacket
// based on the `Layrz Protocol v2` specification
//
// Returns an error if the packet is invalid
func (p *PmPacket) FromPacket(raw *string) error {
	if !strings.HasPrefix(*raw, "<Pm>") || !strings.HasSuffix(*raw, "</Pm>") {
		return errors.New("invalid package, should be <Pm>...</Pm>")
	}

	*raw = strings.TrimPrefix(*raw, "<Pm>")
	*raw = strings.TrimSuffix(*raw, "</Pm>")

	rawCrc := (*raw)[len(*raw)-4:]
	*raw = (*raw)[:len(*raw)-4]

	receivedCrc, err := strconv.ParseUint(rawCrc, 16, 16)
	if err != nil {
		return errors.New("cannot convert CRC to integer")
	}

	calculatedCrc := calculateCrc([]byte(*raw))

	if calculatedCrc != uint16(receivedCrc) {
		return fmt.Errorf("invalid CRC, received: %04X, calculated: %04X", receivedCrc, calculatedCrc)
	}

	*raw = strings.TrimSuffix(*raw, ";")
	parts := strings.Split(*raw, ";")
	if len(parts) != 3 {
		return errors.New("invalid package, should contain 3 parts")
	}

	p.Filename = &parts[0]
	p.ContentType = &parts[1]

	data, err := base64.StdEncoding.DecodeString(parts[2])
	if err != nil {
		return errors.New("cannot decode base64 data")
	}
	p.Data = &data

	return nil
}

// ToPacket is a method that converts a PaPacket to a raw packet
// based on the `Layrz Protocol v2` specification
func (p *PmPacket) ToPacket() *string {
	content := ""
	content += *p.Filename + ";"
	content += *p.ContentType + ";"
	content += base64.StdEncoding.EncodeToString(*p.Data) + ";"

	crc := calculateCrc([]byte(content))
	content += fmt.Sprintf("%04X", crc)

	content = fmt.Sprintf("<Pm>%s</Pm>", content)
	return &content
}

File: dart/lib/src/packets/src/client/pm.dart
part of '../../packets.dart';

class PmPacket extends ClientPacket {
  /// [filename] is the name of the file.
  String filename;

  /// [contentType] is the content type of the file.
  String contentType;

  /// [data] is the content of the file.
  Uint8List data;

  /// [PmPacket] is the media package.
  ///
  /// This package is part of the package sent from the server to the device.
  PmPacket({
    required this.filename,
    required this.contentType,
    required this.data,
  }) : super();

  /// [fromPacket] creates a [PmPacket] from a string package in the format of `Layrz Protocol v3`.
  static PmPacket fromPacket(String raw) {
    if (!raw.startsWith('<Pm>') || !raw.endsWith('</Pm>')) {
      throw ParseException('Invalid identification package, should be <Pm>...</Pm>');
    }

    final parts = raw.substring(4, raw.length - 5).split(';');
    if (parts.length != 4) {
      throw MalformedException('Invalid package parts, should have 4 parts');
    }

    int? receivedCrc = int.tryParse(parts[3], radix: 16);
    int? calculatedCrc = calculateCrc("${parts.sublist(0, 1).join(';')};".codeUnits);

    if (receivedCrc != calculatedCrc) {
      throw CrcException('Invalid CRC, received: $receivedCrc, calculated: $calculatedCrc');
    }

    String filename = parts[0];
    String contentType = parts[1];
    Uint8List data;
    try {
      data = base64Decode(parts[2]);
    } catch (e) {
      throw ParseException('Invalid base64 data');
    }

    return PmPacket(
      filename: filename,
      contentType: contentType,
      data: data,
    );
  }

  /// [toPacket] returns the package in the format of `Layrz Protocol v3`.
  @override
  String toPacket() {
    String payload = '$filename;$contentType;${base64Encode(data)};';
    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();
    return '<Pm>$payload$crc</Pm>';
  }

  @override
  String toString() => toPacket();

  @override
  PmPacket copyWith({
    String? filename,
    String? contentType,
    Uint8List? data,
  }) {
    return PmPacket(
      filename: filename ?? this.filename,
      contentType: contentType ?? this.contentType,
      data: data ?? this.data,
    );
  }
}
